app/src/main/java/org/wikipedia/richtext/CustomHtmlParser.kt

```diff
@@ -0,0 +1,147 @@
+        wrapped?.startPrefixMapping(prefix, uri)
+    }
+
+    override fun endPrefixMapping(prefix: String?) {
+        wrapped?.endPrefixMapping(prefix)
+    }
+
+    override fun characters(ch: CharArray?, start: Int, length: Int) {
+        wrapped?.characters(ch, start, length)
+    }
+
+    override fun ignorableWhitespace(ch: CharArray?, start: Int, length: Int) {
+        wrapped?.ignorableWhitespace(ch, start, length)
+    }
+
+    override fun processingInstruction(target: String?, data: String?) {
+        wrapped?.processingInstruction(target, data)
+    }
+
+    override fun skippedEntity(name: String?) {
+        wrapped?.skippedEntity(name)
+    }
+
+    class CustomTagHandler(private val view: TextView?) : TagHandler {
+        private var lastAClass = ""
+
+        override fun handleTag(opening: Boolean, tag: String?, output: Editable?, attributes: Attributes?): Boolean {
+            if (tag == "img") {
+                return true
+            } else if (tag == "a") {
+                if (opening) {
+                    lastAClass = getValue(attributes, "class").orEmpty()
+                } else if (output != null && output.isNotEmpty()) {
+                    val spans = output.getSpans<URLSpan>(output.length - 1)
+                    if (spans.isNotEmpty()) {
+                        val span = spans.last()
+                        val start = output.getSpanStart(span)
+                        val end = output.getSpanEnd(span)
+                        output.removeSpan(span)
+                        val color = if (lastAClass == "new" && view != null) ResourceUtil.getThemedColor(view.context, R.attr.colorError) else -1
+                        output.setSpan(URLSpanNoUnderline(span.url, color), start, end, 0)
+                    }
+                }
+            }
+            return false
+        }
+    }
+
+    companion object {
+
+        fun fromHtml(html: String?, view: TextView? = null): Spanned {
+            var sourceStr = html.orEmpty()
+
+            if ("<" !in sourceStr && "&" !in sourceStr) {
+                // If the string doesn't contain any hints of HTML entities, then skip the expensive
+                // processing that fromHtml() performs.
+                return sourceStr.toSpanned()
+            }
+
+            sourceStr = sourceStr.replace("&#8206;", "\u200E")
+                .replace("&#8207;", "\u200F")
+                .replace("&amp;", "&")
+
+            // TODO: Investigate if it's necessary to inject a dummy tag at the beginning of the
+            // text, since there are reports that XmlReader ignores the first tag by default?
+            // This would become something like "<inject/>$sourceStr".parseAsHtml(...)
+            return sourceStr.parseAsHtml(HtmlCompat.FROM_HTML_MODE_LEGACY, null,
+                CustomHtmlParser(CustomTagHandler(view)))
+        }
+
+        private fun getValue(attributes: Attributes?, name: String): String? {
```

Comment:
Should we add `@Suppress("SameParameterValue")` for the `name`

Reply:
In the next PR, this will be used with other parameters.